### DISCUSSION TITLE
请教R语言小数点的问题

### DISCUSSION THREAD

**Discussion by Renmr488** [ShixiangWang/self-study#98] [Q&A] [answered]
模型得出的值应该为一个小数，但是出来的是double类型负数。请问如何可以转换为小数呢？
![1653204347(1)](https://user-images.githubusercontent.com/62462250/169683826-2f59f6a5-90ff-4299-893e-c06524fdb7ba.jpg)

**Comment by ShixiangWang** [comment, marked as answer]
@Renmr488 这个问题有点奇怪，负数也是小数，另外R表示浮点数就是double类型。

**Reply by Renmr488** [reply]
我想知道具体的小数值，应该怎么转换呢？

**Reply by ShixiangWang** [reply]
@Renmr488 哦哦，你是说想提取第二列的具体数值？参考下面的代码：

```r
> df = dplyr::tibble(
+   a = c(list(1), list(2), list(3))
+ )
> df
# A tibble: 3 × 1
  a        
  <list>   
1 <dbl [1]>
2 <dbl [1]>
3 <dbl [1]>
> sapply(df$a, function(x) x[1])
[1] 1 2 3
```

**Reply by Renmr488** [reply]
可是提取出来还是负数，为什么呢

> df = dplyr::tibble(
+      a = c(list(meta1[["TE.random"]]), list(meta1[["lower.random"]]), list(meta1[["upper.random"]]))
+    )
> 
> sapply(df$a, function(x) x[1])
[1] -7.271200152 -7.434889543 -7.107510761

**Reply by ShixiangWang** [reply]
@Renmr488 ???

你这个很奇怪。首先我提供的示例是想说明你应该把你截图的第2列传入处理，即`sapply(df[[2]], function(x) x[1])`（`df`代表你的数据）。其次，提取出来的是负数，这个我无法帮你解释，你的dataframe存储的是什么，提取出来的就是什么，负数不是小数？我无法理解。

**Reply by Renmr488** [reply]
模型得出的结果是0.0007（图1），但是我还想知道后面几位小数，多保留几位。然后点开meta1（图2），对应的值应该是TE.random，这个值应该是0.0007，但不知道是多少。
![image](https://user-images.githubusercontent.com/62462250/169687845-7fb4d5d3-083d-49e6-9b8b-532f504597a8.png)
![image](https://user-images.githubusercontent.com/62462250/169687898-2a3ecf63-e4cd-401e-baf9-dbb590af47b4.png)

**Reply by Renmr488** [reply]
我的意思是这个负数-7.271200152可以转换为0.0007吗？

**Reply by ShixiangWang** [reply]
> 我的意思是这个负数-7.271200152可以转换为0.0007吗？

不能。如果我没猜错，你提取出来的和你想象中的不一样。没有可重复的代码，我也无法不知道问题该怎么探索。

**Reply by Renmr488** [reply]
打扰了谢谢。我换了个模型方法后，结果对了。
